aten/src/ATen/native/NativeFunctions.cpp

```diff
@@ -0,0 +1,325 @@
+  dim = maybe_wrap_dim(dim, self.dim() + 1);
+
+  auto g = inferUnsqueezeGeometry(self, dim);
+  return self.as_strided(std::get<0>(g), std::get<1>(g));
+}
+
+Tensor & unsqueeze_(Tensor& self, int64_t dim) {
+  dim = maybe_wrap_dim(dim, self.dim() + 1);
+
+  auto g = inferUnsqueezeGeometry(self, dim);
+  return self.as_strided_(std::get<0>(g), std::get<1>(g));
+}
+
+Tensor stack(TensorList tensors, int64_t dim) {
+  if (tensors.size() == 0) {
+    throw std::runtime_error("stack expects a non-empty TensorList");
+  }
+  dim = maybe_wrap_dim(dim, tensors[0].dim() + 1);
+
+  std::vector<Tensor> inputs(tensors.size());
+  for (size_t i = 0; i < tensors.size(); ++i) {
+    inputs[i] = tensors[i].unsqueeze(dim);
+  }
+  return at::cat(inputs, dim);
+}
+
+std::tuple<at::Tensor, at::Tensor> SpatialRoIPooling_forward(
+	const Tensor& input,
+	const Tensor& rois,
+	int64_t pooledHeight,
+	int64_t pooledWidth,
+	double spatialScale)
+{
+	// Input is the output of the last convolutional layer in the Backbone network, so
+	// it should be in the format of NCHW
+	AT_ASSERT(input.ndimension() == 4, "Input to RoI Pooling should be a NCHW Tensor");
+
+	// ROIs is the set of region proposals to process. It is a 2D Tensor where the first
+	// dim is the # of proposals, and the second dim is the proposal itself in the form
+	// [batch_index startW startH endW endH]
+	AT_ASSERT(rois.ndimension() == 2, "RoI Proposals should be a 2D Tensor, (batch_sz x proposals)");
+	AT_ASSERT(rois.size(1) == 5, "Proposals should be of the form [batch_index startW startH endW enH]");
+
+	auto proposals = rois.size(0);
+	auto inputChannels = input.size(1);
+	auto inputHeight = input.size(2);
+	auto inputWidth = input.size(3);
+
+	// Output Tensor is (num_rois, C, pooledHeight, pooledWidth)
+	auto output = input.type().tensor({proposals, inputChannels, pooledHeight, pooledWidth});
+
+	// TODO: need some mechanism for determining train vs. test
+
+	// During training, we need to store the argmaxes for the pooling operation, so
+	// the argmaxes Tensor should be the same size as the output Tensor
+	auto argmaxes = input.type().toScalarType(kInt).tensor({proposals, inputChannels, pooledHeight, pooledWidth});
+
+	AT_ASSERT(input.is_contiguous(), "input must be contiguous");
+	AT_ASSERT(rois.is_contiguous(), "rois must be contiguous");
+
+	auto *rawInput = input.data<float>();
+	auto inputChannelStride = inputHeight * inputWidth;
+	auto inputBatchStride = inputChannels * inputChannelStride;
+	auto *rawRois = rois.data<float>();
+	auto roiProposalStride = rois.size(1);
+
+	auto *rawOutput = output.data<float>();
+	auto *rawArgmaxes = argmaxes.data<int>();
+	auto outputChannelStride = pooledHeight * pooledWidth;
+
+	// Now that our Tensors are properly sized, we can perform the pooling operation.
+	// We iterate over each RoI and perform pooling on each channel in the input, to
+	// generate a pooledHeight x pooledWidth output for each RoI
+	for (auto i = 0; i < proposals; ++i) {
```

Comment:
I should probably fix this to use spaces instead of tabs, which is what it appears to be doing.